crates/gitbutler-branch-actions/src/branch_manager/branch_removal.rs

```diff
@@ -95,51 +96,100 @@ impl BranchManager<'_> {
-                            .collect::<Vec<(PathBuf, Vec<VirtualBranchHunk>)>>();
-                        let tree_oid =
-                            gitbutler_diff::write::hunks_onto_oid(self.ctx, stack.head(), files)?;
-                        let mut merge = gix_repo.merge_trees(
-                            git2_to_gix_object_id(base_tree_id),
-                            final_tree_id,
-                            git2_to_gix_object_id(tree_oid),
-                            gix_repo.default_merge_labels(),
-                            merge_options.clone(),
-                        )?;
-                        let final_tree_id = merge.tree.write()?.detach();
-                        Ok(final_tree_id)
-                    },
-                )?;
-            repo.find_tree(gix_to_git2_oid(final_tree_id))?
-        };
+        if self.ctx.app_settings().feature_flags.v3 {
+            // On v3 we want to take the `current_wd_tree` and try to extract
+            // whatever branch we want to unapply. There are a handful of ways
+            // to achieve this, including calculating the inverse diff and
+            // applying that.
+            //
+            // We can however do more or less what `git revert` does, and
+            // perform a three way merge where the `ours` side is the cwdt, the
+            // `theirs` side is the workspace root, and the `base` is the head
+            // of the branch we want to unapply.
+            //
+            // In order to handle locked files, I'm going to choose to
+            // resolve conflicts in the favor of `ours` (the cwdt) which will
+            // keep any locked changes in the cwdt.
 
-        let _span = tracing::debug_span!("checkout final tree").entered();
-        // checkout final_tree into the working directory
-        repo.checkout_tree_builder(&final_tree)
-            .force()
-            .remove_untracked()
-            .checkout()
-            .context("failed to checkout tree")?;
+            let gix_repo = self.ctx.gix_repository()?;
+            let merge_options = gix_repo
+                .tree_merge_options()?
+                .with_file_favor(Some(gix::merge::tree::FileFavor::Ours))
+                .with_tree_favor(Some(gix::merge::tree::TreeFavor::Ours));
+
+            let cwdt = repo.create_wd_tree(0)?.id().to_gix();
+            let workspace_base = gix_repo
+                .find_commit(workspace_base(self.ctx, perm.read_permission())?)?
+                .tree_id()?;
+            let stack_head = gix_repo.find_commit(stack.head().to_gix())?.tree_id()?;
+
+            let mut merge = gix_repo.merge_trees(
+                stack_head,
+                cwdt,
+                workspace_base,
+                gix_repo.default_merge_labels(),
+                merge_options,
+            )?;
+            let tree = merge.tree.write()?;
+            let tree = repo.find_tree(tree.object()?.id().to_git2())?;
```

Comment:
`tree` is already a tree-id, so there should be no need for fetching its `object()` beforehand.